File: clients/php/Debugging/DebugCommand.php
<?php

namespace Debugging {
    require_once 'Debugging/DebugData.php';
    require_once 'Stream.php';

    /**
     * Debug commands that can be sent while debugging with the app
     */
    abstract class DebugCommand
    {
        /**
         * Write DebugCommand to output stream
         */
        abstract function writeTo(\OutputStream $stream): void;

        /**
         * Read DebugCommand from input stream
         */
        static function readFrom(\InputStream $stream): DebugCommand
        {
            $tag = $stream->readInt32();
            if ($tag == \Debugging\DebugCommand\Add::TAG) {
                return \Debugging\DebugCommand\Add::readFrom($stream);
            }
            if ($tag == \Debugging\DebugCommand\Clear::TAG) {
                return \Debugging\DebugCommand\Clear::readFrom($stream);
            }
            if ($tag == \Debugging\DebugCommand\SetAutoFlush::TAG) {
                return \Debugging\DebugCommand\SetAutoFlush::readFrom($stream);
            }
            if ($tag == \Debugging\DebugCommand\Flush::TAG) {
                return \Debugging\DebugCommand\Flush::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace Debugging\DebugCommand {
    /**
     * Add debug data to current tick
     */
    class Add extends \Debugging\DebugCommand
    {
        const TAG = 0;
    
        /**
         * Data to add
         */
        public \Debugging\DebugData $debugData;
    
        function __construct(\Debugging\DebugData $debugData)
        {
            $this->debugData = $debugData;
        }
    
        /**
         * Read Add from input stream
         */
        public static function readFrom(\InputStream $stream): Add
        {
            $debugData = \Debugging\DebugData::readFrom($stream);
            return new Add($debugData);
        }
        
        /**
         * Write Add to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(Add::TAG);
            $this->debugData->writeTo($stream);
        }
    }

    /**
     * Clear current tick's debug data
     */
    class Clear extends \Debugging\DebugCommand
    {
        const TAG = 1;
    
    
        function __construct()
        {
        }
    
        /**
         * Read Clear from input stream
         */
        public static function readFrom(\InputStream $stream): Clear
        {
            return new Clear();
        }
        
        /**
         * Write Clear to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(Clear::TAG);
        }
    }

    /**
     * Enable/disable auto performing of commands
     */
    class SetAutoFlush extends \Debugging\DebugCommand
    {
        const TAG = 2;
    
        /**
         * Enable/disable autoflush
         */
        public bool $enable;
    
        function __construct(bool $enable)
        {
            $this->enable = $enable;
        }
    
        /**
         * Read SetAutoFlush from input stream
         */
        public static function readFrom(\InputStream $stream): SetAutoFlush
        {
            $enable = $stream->readBool();
            return new SetAutoFlush($enable);
        }
        
        /**
         * Write SetAutoFlush to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(SetAutoFlush::TAG);
            $stream->writeBool($this->enable);
        }
    }

    /**
     * Perform all previously sent commands
     */
    class Flush extends \Debugging\DebugCommand
    {
        const TAG = 3;
    
    
        function __construct()
        {
        }
    
        /**
         * Read Flush from input stream
         */
        public static function readFrom(\InputStream $stream): Flush
        {
            return new Flush();
        }
        
        /**
         * Write Flush to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(Flush::TAG);
        }
    }
}

File: clients/rust/src/debugging/debug_command.rs
use super::*;

/// Debug commands that can be sent while debugging with the app
#[derive(Clone, Debug)]
pub enum DebugCommand {
    /// Add debug data to current tick
    Add {
        /// Data to add
        debug_data: debugging::DebugData,
    },
    /// Clear current tick's debug data
    Clear {
    },
    /// Enable/disable auto performing of commands
    SetAutoFlush {
        /// Enable/disable autoflush
        enable: bool,
    },
    /// Perform all previously sent commands
    Flush {
    },
}

impl trans::Trans for DebugCommand {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Self::Add {
                debug_data,
            } => {
                <i32 as trans::Trans>::write_to(&0, writer)?;
                debug_data.write_to(writer)?;
            }
            Self::Clear {
            } => {
                <i32 as trans::Trans>::write_to(&1, writer)?;
            }
            Self::SetAutoFlush {
                enable,
            } => {
                <i32 as trans::Trans>::write_to(&2, writer)?;
                enable.write_to(writer)?;
            }
            Self::Flush {
            } => {
                <i32 as trans::Trans>::write_to(&3, writer)?;
            }
        }
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => {
                let debug_data: debugging::DebugData = trans::Trans::read_from(reader)?;
                Ok(Self::Add {
                    debug_data,
                })
            }
            1 => {
                Ok(Self::Clear {
                })
            }
            2 => {
                let enable: bool = trans::Trans::read_from(reader)?;
                Ok(Self::SetAutoFlush {
                    enable,
                })
            }
            3 => {
                Ok(Self::Flush {
                })
            }
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}